feat: add the ability to copy module import to the clipboard

hadzhiyski opened the pull request:
Closes #628

Related changes:
- https://github.com/angular/components/pull/19200

hadzhiyski committed:
feat: add the ability to copy module import to the clipboard

Splaktar commented on src/app/shared/doc-viewer/docs-api-module-import/docs-api-module-import.ts:
```diff
@@ -0,0 +1,50 @@
+      class="docs-api-module-import-copy"
+      mat-icon-button
+      (click)="copyModuleImport()"
+      matTooltip="Copy module import">
```
We need to have an `aria-label` here for a11y.

Splaktar commented on src/app/shared/doc-viewer/docs-api-module-import/docs-api-module-import.ts:
```diff
@@ -0,0 +1,50 @@
+import { Component, Input, AfterViewInit, OnInit } from "@angular/core";
+import { MatSnackBar } from "@angular/material/snack-bar";
+import { CopierService } from "../../copier/copier.service";
```
We should probably look at replacing the `CopierService` with the new https://material.angular.io/cdk/clipboard/overview.

hadzhiyski replied:
I was wondering about the `CopierService`, but left it that way for consistency. There are other components which are using it as well, for example the `ExampleViewer` component. It might be a good idea to switch to `Clipboard` with separate PR.

Splaktar reviewed (changes requested):
Thank you for the contribution!

Can you please post a screenshot in the OP?

Splaktar commented on src/app/shared/doc-viewer/docs-api-module-import/docs-api-module-import.ts:
```diff
@@ -0,0 +1,50 @@
+
+  copyModuleImport(): void {
+    if (this.copier.copyText(this.moduleImport)) {
+      this.snackbar.open("Module import copied", "", { duration: 2500 });
```
Please run `npm run lint` and correct the issues with double quotes. You may be able to auto correct some of them via `ng lint --fix`.

hadzhiyski replied:
haven't noticed the double quotes. well, that's a shame on my side

Splaktar commented on src/app/shared/doc-viewer/docs-api-module-import/docs-api-module-import.ts:
```diff
@@ -0,0 +1,50 @@
+const MODULE_IMPORT_TEMPLATE = `import ${EXPORT_NAME_PLACEHOLDER} from '${MODULE_IMPORT_PATH_PLACEHOLDER}';`;
+
+@Component({
+  selector: "docs-api-module-import",
```
Based on PR https://github.com/angular/components/pull/19200, should this be `'[docs-api-module-import]'`?

hadzhiyski committed:
chore: lint errors fix and a11y description

hadzhiyski commented:
![image](https://user-images.githubusercontent.com/20074286/80682004-6625c680-8aca-11ea-9fe7-712017865cc5.png)

Here is a screenshot of the changes. Ignore the fact that `Material autocomplete` is importing `MatButtonModule`. It is just for the test.

hadzhiyski committed:
chore: add curly braces around export name.

Reason for change: https://github.com/angular/components/pull/19200#discussion_r419101641

hadzhiyski commented:
@Splaktar this PR is ready for review.

Splaktar commented on angular.json:
```diff
@@ -130,24 +134,28 @@
+                "bundleName": "assets/indigo-pink"
               },
               {
+                "lazy": true,
```
`lazy` is the old deprecated name for `inject`, we don't need both.

Can you please try rebasing your branch on `master` as well?

Splaktar commented on src/app/shared/doc-viewer/docs-api-module-import/docs-api-module-import.ts:
```diff
@@ -0,0 +1,51 @@
+      mat-icon-button
+      (click)="copyModuleImport()"
+      matTooltip="Copy module import">
+      <span class="material-icons">file_copy</span>
```
```suggestion
      <span class="material-icons">content_copy</span>
```
This is the icon that angular.io uses.

Splaktar commented on src/app/shared/doc-viewer/docs-api-module-import/docs-api-module-import.ts:
```diff
@@ -0,0 +1,51 @@
+      (click)="copyModuleImport()"
+      matTooltip="Copy module import">
+      <span class="material-icons">file_copy</span>
+    </button>
```
We could add some opacity to the icon as it's a bit too strong in pure black. It should still meet a11y contrast guidelines though (in light and dark modes). 

Angular.io changes the color to a darker color on hover rather than changing opacity. So we should consider following that pattern.

hadzhiyski replied:
I added a button color style, the same way it is applied for `example-viewer` and the global icon color styles which are for H3 and H4.

Splaktar reviewed (changes requested):
Thank you for making those updates. Sorry for the delay in getting to this review.

Something happened with your last update/rebase. I don't think that this change should have 49 changed files and many of them seem to be coming from other commits. Can you please rebase your commit on top of the latest `master` and force push it up again?

hadzhiyski committed:
Update src/app/shared/doc-viewer/docs-api-module-import/docs-api-module-import.ts

Co-authored-by: Michael Prentice <splaktar@gmail.com>

hadzhiyski committed:
fix: Add opacity to copy button

hadzhiyski commented:
Hello @Splaktar,
It has been a while since your last review. From my perspective, everything is covered. Do you expect something else which I might have missed?

Splaktar reviewed (changes requested):
I checked this out locally to test it. I also rebased it on the latest `master`. However, I was unable to get the copy icon to appear or be rendered?

![Screen Shot 2020-07-14 at 16 31 52](https://user-images.githubusercontent.com/3506071/87473739-f49ce500-c5ef-11ea-979d-e249b1eb0ab4.png)

![Screen Shot 2020-07-14 at 16 35 42](https://user-images.githubusercontent.com/3506071/87473826-15fdd100-c5f0-11ea-91f0-265fb2fb071a.png)

hadzhiyski commented:
Hi,
Most probably the reason is this https://github.com/angular/components/pull/19200 . If you test it locally, you should rebuild the docs locally as well, because their PR is still open.

Splaktar commented:
OK, the other PR looks like it's ready to go, hopefully we can merge that soon. Can you please rebase this PR as well?

hadzhiyski commented:
@googlebot I consent.

hadzhiyski commented:
@googlebot I consent
